python/triton/language/semantic.py

```diff
@@ -1677,6 +1677,31 @@ def associative_scan(inputs: Sequence[tl.tensor], axis: int, region_builder_fn,
     return tuple(wrap_tensor(scan_op.get_result(i), inputs[i].type.scalar, shape) for i in range(len(inputs)))
 
 
+# ===----------------------------------------------------------------------===
+#                               Gather
+# ===----------------------------------------------------------------------===
+
+
+def gather(src: tl.tensor, index: tl.tensor, axis: int, builder: ir.builder) -> tl.tensor:
+    assert index.dtype.is_int(), "index must be an integer tensor"
+
+    rank = len(src.type.shape)
+    assert len(index.type.shape) == rank, "source and index tensors must have the same rank"
```

Comment:
Would be nice to support broadcasting.

Reply:
Can you elaborate on what the broadcasting semantics would be?

Reply:
https://pytorch.org/docs/stable/notes/broadcasting.html

Reply:
Actually, PyTorch does not mplements broadcasting for this operation.
If you think about it, it's a bit weird doing something like `gather(3d_tensor, dim=1, 2d_tensor)`.
I say we don't implement broadcast semantics just yet. We can always implement them later.

Reply:
@peterbell10 What are you thoughts here?